src/net/socket.rs

```diff
@@ -131,34 +123,52 @@ impl Socket {
 
     // Serializes and sends a `Packet` on the socket. On success, returns the number of bytes written.
     fn send_to(&mut self, packet: Packet) -> Result<usize> {
-        let connection = self
-            .connections
-            .get_or_insert_connection(packet.addr(), &self.config);
-        let mut packet_data =
-            connection.process_outgoing(packet.payload(), packet.delivery_method())?;
+        let (dropped_packets, processed_packet) = {
+            let connection = self
+                .connections
+                .get_or_insert_connection(packet.addr(), &self.config);
+
+            let processed_packet = connection.process_outgoing(
+                packet.payload(),
+                packet.delivery_guarantee(),
+                packet.order_guarantee(),
+            )?;
+
+            (connection.gather_dropped_packets(), processed_packet)
+        };
+
         let mut bytes_sent = 0;
 
-        if let Some(link_conditioner) = &self.link_conditioner {
-            if link_conditioner.should_send() {
-                for payload in packet_data.parts() {
-                    bytes_sent += self.send_packet(&packet.addr(), &payload)?;
-                }
-            }
+        let should_send = if let Some(link_conditioner) = &self.link_conditioner {
+            link_conditioner.should_send()
         } else {
-            for payload in connection.gather_dropped_packets() {
-                bytes_sent += self.send_packet(&packet.addr(), &payload)?;
+            true
+        };
+
+        if should_send {
+            match processed_packet {
+                Outgoing::Packet(outgoing) => {
+                    bytes_sent += self.send_packet(&packet.addr(), &outgoing.contents())?;
+                }
+                Outgoing::Fragments(packets) => {
+                    for outgoing in packets {
+                        bytes_sent += self.send_packet(&packet.addr(), &outgoing.contents())?;
+                    }
+                }
             }
 
-            for payload in packet_data.parts() {
+            for payload in dropped_packets {
                 bytes_sent += self.send_packet(&packet.addr(), &payload)?;
             }
+
+            return Ok(bytes_sent);
         }
 
-        Ok(bytes_sent)
+        Ok(0)
     }
 
     // Receives a single message from the socket. On success, returns the packet containing origin and data.
```

Comment:
Since this isn't returning the packet anymore, we should probably update the comment